src/tools/perf/lib/libperf.c

```diff
@@ -1520,26 +1520,32 @@ static ucs_status_t ucp_perf_test_flush_workers(ucx_perf_context_t *perf)
 
 static ucs_status_t ucp_perf_test_setup_self_endpoints(ucx_perf_context_t *perf)
 {
-    ucp_worker_h worker = perf->ucp.tctx[0].perf.ucp.worker;
+    ucp_worker_attr_t worker_attr = {
+        .field_mask = UCP_WORKER_ATTR_FIELD_ADDRESS,
+        .address    = NULL
+    };
+    ucp_worker_h worker;
     ucs_status_t status;
     ucp_ep_params_t ep_params;
     ucx_perf_context_t *thread_perf;
-    ucp_worker_attr_t worker_attr = {
-        .field_mask = UCP_WORKER_ATTR_FIELD_ADDRESS
-    };
     unsigned i;
 
-    status = UCX_PERF_VERBOSE(error, &perf->params, ucp_worker_query, worker,
-                              &worker_attr);
-    if (status != UCS_OK) {
-        return status;
-    }
-
-    ep_params.field_mask = UCP_EP_PARAM_FIELD_REMOTE_ADDRESS;
-    ep_params.address    = worker_attr.address;
-
     for (i = 0; i < perf->params.thread_count; ++i) {
         thread_perf = &perf->ucp.tctx[i].perf;
+        worker      = thread_perf->ucp.worker;
+        /* coverity[use_after_free] */
+        status      = UCX_PERF_VERBOSE(error, &perf->params, ucp_worker_query,
+                                       worker, &worker_attr);
```

Comment:
So the problem here was that ep from all threads were connected to the "main" worker and not the "thread" worker? How is this fix related to MEMIC?

Reply:
maybe it has to be a separate fix in another PR?
does it have some dependency with MEMIC fix?

Reply:
We can split it but we won't be able to test memic with multithreading using just one of those fixes, what about separating this PR into 2 different commits when squashing?

Reply:
Will make this PR the memic fix and open a new PR for the perftest

Reply:
https://github.com/openucx/ucx/pull/9766